app/src/main/java/com/ataulm/artcollector/gallery/data/AndroidGalleryRepository.kt

```diff
@@ -1,21 +1,33 @@
 package com.ataulm.artcollector.gallery.data
 
+import androidx.paging.PagingSource
 import com.ataulm.artcollector.ApiObjectRecord
+import com.ataulm.artcollector.Artist
 import com.ataulm.artcollector.HarvardArtMuseumApi
-import com.ataulm.artcollector.domain.Artist
-import com.ataulm.artcollector.domain.Painting
-import com.ataulm.artcollector.domain.Gallery
-import com.ataulm.artcollector.gallery.domain.GalleryRepository
+import com.ataulm.artcollector.Painting
+import com.ataulm.artcollector.datanotdomain.GalleryRepository
 import javax.inject.Inject
 
 internal class AndroidGalleryRepository @Inject constructor(
         private val harvardArtMuseumApi: HarvardArtMuseumApi
 ) : GalleryRepository {
 
-    override suspend fun gallery(): Gallery {
-        val paintings = harvardArtMuseumApi.gallery().await().records
-                .map { it.toPainting() }
-        return Gallery(paintings)
+    override fun gallery(): PagingSource<Int, Painting> {
+        return GalleryPagingSource(harvardArtMuseumApi)
+    }
+}
+
+private class GalleryPagingSource(private val harvardArtMuseumApi: HarvardArtMuseumApi)
+    : PagingSource<Int, Painting>() {
+
+    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, Painting> {
+        val result = harvardArtMuseumApi.gallery(page = params.key, pageSize = params.loadSize)
+        // TODO: handle errors
+        return LoadResult.Page(
+                data = result.records.map { it.toPainting() },
+                prevKey = if (result.info.page == 1) null else result.info.page - 1,
```

Comment:
If we've just loaded the first page, then there is no previous page (null). Otherwise the previous key is "this page - 1".

Same but opposite for next key 👇